### ISSUE TITLE
Bounding Boxes

### ISSUE THREAD

**Issue by ffreyer** [MakieOrg/Makie.jl#1653]
I wanted to collect some issues/thoughts on bounding boxes. I'll work on them after #1596 gets merged.

#### Bounding Boxes ignore marker sizes

They probably shouldn't...

```julia
fig, ax, p = meshscatter(Point2f[(0,0), (1,1)], marker = Rect2f(-0.5, -0.5, 1, 1))
bb = boundingbox(p)
# GeometryBasics.HyperRectangle{3, Float32}(Float32[0.0, 0.0, 0.0], Float32[1.0, 1.0, 0.0])
lines!(ax, Rect2f(bb), color = :red)
```

![Screenshot from 2022-02-08 15-53-52](https://user-images.githubusercontent.com/10947937/153013306-304bb2a5-9883-4161-a7d3-752c749b54de.png)

#### User facing text plots have no bounding box

```julia
fig, ax, p = text("Test")
bb = boundingbox(p)
# GeometryBasics.HyperRectangle{3, Float32}(Float32[0.0, 0.0, 0.0], Float32[0.0, 0.0, 0.0])
```

#### Internal text bounding boxes always have pixel units widths and data space origins

They should be in whatever space the primary plot is in 

```julia
fig, ax, p = text("Test", position = Point2f(7.3, -0.3))
bb = boundingbox(p.plots[1])
# GeometryBasics.HyperRectangle{3, Float32}(Float32[7.3, -0.3, 0.0], Float32[42.36328, 23.28125, 0.0])
```

#### Bounding boxes in different spaces

Sometimes it may be useful to get a bounding box in a different space than the plot is in. For example, if you want to annotate a point in 3D with a white background pane it would be useful to plot the text at a 3D data space position and the pane using its screen/pixel space bounding box.

#### Per-marker bounding boxes

Getting bounding boxes per marker or whatever makes sense for a given plot might also be useful. I think it would be necessary for a ray based point picking system and probably for a collision system. It can also be useful to highlight hovered objects with `pick`.

**Comment by koehlerson** [comment]
Let me know if I should open a separate issue for it, but it seems that the `space=:pixel` option doesn't work in recipes 
```julia
t = Makie.text!(Ele,"$id", position=position, textsize=Ele[:textsize])
bb = Makie.boundingbox(t)
Makie.wireframe!(Ele,bb,space=:pixel)
```
produces 
![grafik](https://user-images.githubusercontent.com/27497290/196181385-ce6ebb53-d79e-4087-9a05-3906fca16389.png)

**Comment by ffreyer** [comment]
That's a separate issue. What's happening in your case is that bounding boxes are calculated with the limits from just after `text!(...)`. At that point they should be correct. Once you add the wireframe plot however the limits get updated and thus the bounding boxes are no longer correct. The fix (in Makie) is to ignore anything but `space = :data` from triggering limit updates like it's done for Axis3 and Axis.

**Comment by ffreyer** [comment]
I can't reproduce this with
```julia
fig = Figure()
ax = LScene(fig[1, 1])
t = text!(ax, ["test 1", "test 2"], position = [Point3f(0), Point3f(1)])
bb = boundingbox(t)
wireframe!(ax, bb, space = :pixel)
fig
```
and the limit updates seem to all go through `center!` which already excludes `space = :pixel`. Maybe your version is too old?

**Comment by koehlerson** [comment]
It works outside of recipes for me as well! I observe this behavior only within a recipe

**Comment by jkrumbiegel** [comment]
I think Makie still has to clean up some logic with the limits of nested plots. There's also a related problem if you want to use `vlines!` inside a plot recipe, its `yautolimits = false` attribute doesn't work anymore.

**Comment by ffreyer** [comment]
Ah, right, that's because 
https://github.com/MakieOrg/Makie.jl/blob/ae280552a97596c6beedbdc59d3c246343ddb277/src/layouting/data_limits.jl#L217-L225
only considers `exclude` for the top level plot. That should be happening with Axis3 and Axis too...

**Comment by jkrumbiegel** [comment]
At least for `xautolimits` etc. a simple `exclude` doesn't work anyway because for `vlines!` as the example, you want to include where the lines are in x, but not in y (because that's scene-projection-dependent). I currently run that logic on the `Axis` level but that's problematic when using recipes.

**Comment by ffreyer** [comment]
I'm closing this in favor of #2881
